spynnaker/pyNN/models/neural_projections/connectors/pool_dense_connector.py

```diff
@@ -0,0 +1,340 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import numpy
+from spinn_utilities.overrides import overrides
+from spinn_front_end_common.utilities.constants import (
+    BYTES_PER_WORD, BYTES_PER_SHORT)
+from pyNN.random import RandomDistribution
+from spynnaker.pyNN.exceptions import SynapticConfigurationException
+from .abstract_connector import AbstractConnector
+from data_specification.enums.data_type import DataType
+from collections.abc import Iterable
+from spinn_front_end_common.utilities.exceptions import ConfigurationException
+from spynnaker.pyNN.models.abstract_models import HasShapeKeyFields
+
+
+_DIMENSION_SIZE = (2 * BYTES_PER_WORD) + (6 * BYTES_PER_SHORT)
+_KEY_INFO_SIZE = 2 * BYTES_PER_WORD
+_CONN_SIZE = _KEY_INFO_SIZE + (2 * BYTES_PER_WORD) + (2 * BYTES_PER_SHORT)
+_DIM_DTYPE = [("mask", "uint32"), ("shift", "uint32"), ("pre_start", "uint16"),
+              ("pre_in_post_start", "uint16"), ("pre_in_post_end", "uint16"),
+              ("pre_in_post_shape", "uint16"), ("recip_pool_stride", "uint16"),
+              ("_PADDING", "uint16")]
+
+
+class PoolDenseConnector(AbstractConnector):
+    """
+    Where the pre- and post-synaptic populations are considered as a 2D\
+    array. Connect every post(row, col) neuron to many pre(row, col, kernel)\
+    through a (kernel) set of weights and/or delays.
+    """
+
+    __slots__ = [
+        "__weights",
+        "__pool_shape",
+        "__pool_stride",
+        "__positive_receptor_type",
+        "__negative_receptor_type"
+    ]
+
+    def __init__(self, weights, pool_shape=None, pool_stride=None,
+                 positive_receptor_type="excitatory",
+                 negative_receptor_type="inhibitory", safe=True,
+                 verbose=False, callback=None):
+        """
+        :param weights:
+            The synaptic strengths
+            Can be:
+            * single value: the same value will be used for all weights
+            * list: the total number of elements must be\
+                    (num after pooling * num post)
+            * numpy.ndarray: As above for list
+            * RandomDistribution: weights will be drawn at random
+        :type weights:
+            int or float or list or numpy.ndarray or RandomDistribution
+        :param pool_shape:
+            Shape of average pooling. If a single value is provided, it will
+            be used for every dimension, otherwise must be the same number of
+            values as there are dimensions in the source.
```

Comment:
I'm not sure this is actually checked anywhere.

Reply:
It is, during self.__to_nd_shape.  It can't be checked before that I don't think as you don't know the dimensions of the pre and post before that.